openhands/integrations/provider.py

```diff
@@ -0,0 +1,138 @@
+from enum import Enum
+
+from pydantic import BaseModel, SecretStr, SerializationInfo, field_serializer
+
+from openhands.integrations.github.github_service import GithubServiceImpl
+from openhands.integrations.gitlab.gitlab_service import GitLabServiceImpl
+from openhands.integrations.service_types import (
+    AuthenticationError,
+    GitService,
+    Repository,
+    User,
+)
+
+
+class ProviderType(Enum):
+    GITHUB = 'github'
+    GITLAB = 'gitlab'
+
+
+class ProviderToken(BaseModel):
+    token: SecretStr | None
+    user_id: str | None
+
+
+PROVIDER_TOKEN_TYPE = dict[ProviderType, ProviderToken]
+CUSTOM_SECRETS_TYPE = dict[str, SecretStr]
+
+
+class SecretStore(BaseModel):
+    provider_tokens: PROVIDER_TOKEN_TYPE = {}
+
+    @classmethod
+    def _convert_token(
+        cls, token_value: str | ProviderToken | SecretStr
+    ) -> ProviderToken:
+        if isinstance(token_value, ProviderToken):
+            return token_value
+        elif isinstance(token_value, str):
+            return ProviderToken(token=SecretStr(token_value), user_id=None)
+        elif isinstance(token_value, SecretStr):
+            return ProviderToken(token=token_value, user_id=None)
+        else:
+            raise ValueError(f'Invalid token type: {type(token_value)}')
+
+    def model_post_init(self, __context) -> None:
+        # Convert any string tokens to ProviderToken objects
+        converted_tokens = {}
+        for token_type, token_value in self.provider_tokens.items():
+            if token_value:  # Only convert non-empty tokens
+                try:
+                    if isinstance(token_type, str):
+                        token_type = ProviderType(token_type)
+                    converted_tokens[token_type] = self._convert_token(token_value)
+                except ValueError:
+                    # Skip invalid provider types or tokens
+                    continue
+        self.provider_tokens = converted_tokens
+
+    @field_serializer('provider_tokens')
+    def provider_tokens_serializer(
+        self, provider_tokens: PROVIDER_TOKEN_TYPE, info: SerializationInfo
+    ):
+        tokens = {}
+        expose_secrets = info.context and info.context.get('expose_secrets', False)
+
+        for token_type, provider_token in provider_tokens.items():
+            if not provider_token or not provider_token.token:
+                continue
+
+            token_type_str = (
+                token_type.value
+                if isinstance(token_type, ProviderType)
+                else str(token_type)
+            )
+            tokens[token_type_str] = {
+                'token': provider_token.token.get_secret_value()
+                if expose_secrets
+                else '**********',
+                'user_id': provider_token.user_id,
+            }
```

Comment:
is this not handled by the use of SecretStr?

Reply:
It's similar to https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/server/settings.py#L52-#L56

I could use `pydantic_encoder(provider_token.token)` instead of `'**********'` for consistency